### DISCUSSION TITLE
How do I implement a LanguageConfiguration?

### DISCUSSION THREAD

**Discussion by miguellaher** [eclipse/tm4e#518] [Q&A] [answered]
I'm trying to add indentation rules inside getOnEnterRules() and use it to my team's project but it seems I need to implement LanguageConfiguration. Any idea how to do this?

**Comment by angelozerr** [comment, marked as answer]
The main idea is to declare on enter rules with declaration mean by using JSON and regexp.

The steps are:

 * create a JSON file like https://github.com/eclipse/wildwebdeveloper/blob/1830af96a8f0a6028df12a6bb2faf20b6d6633cb/org.eclipse.wildwebdeveloper.xml/language-configurations/xml/xml.language-configuration.json#L18
 * register your JSON file with extension point like https://github.com/eclipse/wildwebdeveloper/blob/1830af96a8f0a6028df12a6bb2faf20b6d6633cb/org.eclipse.wildwebdeveloper.xml/plugin.xml#L152

If you don't want to create a plugin for that you can use the UI preferences to register your JSON language configuration file:

![image](https://user-images.githubusercontent.com/1932211/229869386-cd4abaf5-b6c3-45fb-a772-e4dc887c7a0c.png)

**Reply by miguellaher** [reply]
Where can I know that the rule I created is running? Where can I debug this?

**Reply by angelozerr** [reply]
If you see your language configuration JSON mapped with your content type in the UI preferences, it means that it is registered. There is no way to check your regular expression.

**Reply by angelozerr** [reply]
> Where can I debug this?

You want to debug TM4E plugin language configuration?

**Reply by miguellaher** [reply]
> > Where can I debug this?
> 
> You want to debug TM4E plugin language configuration?

Yes. I want to debug the on enter rule that I created. Also, is there any documentation where I can learn how rules function?

**Reply by angelozerr** [reply]
I suggest that you set a breakpoint at https://github.com/eclipse/tm4e/blob/1e29ca71bb2d13131841b25f1a422402b556820f/org.eclipse.tm4e.languageconfiguration/src/main/java/org/eclipse/tm4e/languageconfiguration/internal/registry/LanguageConfigurationRegistryManager.java#L153 and you see what it happens.

Otherwise I suggest that you start with a simple regexp on enter rules.

**Reply by miguellaher** [reply]
Ok. Thanks
